integration/spark/shared/src/main/java/io/openlineage/spark/api/JobNameBuilder.java

```diff
@@ -0,0 +1,126 @@
+/*
+/* Copyright 2018-2024 contributors to the OpenLineage project
+/* SPDX-License-Identifier: Apache-2.0
+*/
+
+package io.openlineage.spark.api;
+
+import static io.openlineage.spark.agent.lifecycle.ExecutionContext.CAMEL_TO_SNAKE_CASE;
+import static io.openlineage.spark.agent.util.DatabricksUtils.prettifyDatabricksJobName;
+
+import io.openlineage.spark.agent.util.DatabricksUtils;
+import io.openlineage.spark.api.SparkOpenLineageConfig.JobNameConfig;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.spark.SparkConf;
+import org.apache.spark.SparkContext;
+import org.apache.spark.sql.execution.SparkPlan;
+import org.apache.spark.sql.execution.WholeStageCodegenExec;
+
+@Slf4j
+public class JobNameBuilder {
+  private static final String JOB_NAME_PARTS_SEPARATOR = ".";
+  private static final String INNER_SEPARATOR = "_";
+
+  public static String build(OpenLineageContext context) {
+    if (context.getJobName() != null) {
+      return context.getJobName();
+    }
+
+    Optional<SparkContext> sparkContext =
+        context.getQueryExecution().map(qe -> qe.executedPlan()).map(SparkPlan::sparkContext);
+    Optional<SparkConf> sparkConf = sparkContext.map(SparkContext::getConf);
+
+    final StringBuilder jobNameBuilder = new StringBuilder();
+    jobNameBuilder.append(
+        normalizeName(
+            Optional.ofNullable(context.getOpenLineageConfig().getOverriddenAppName())
+                .orElse(sparkContext.map(SparkContext::appName).orElse(""))));
+
+    sparkNodeName(context)
+        .ifPresent(
+            nodeName ->
+                jobNameBuilder
+                    .append(JOB_NAME_PARTS_SEPARATOR)
+                    .append(replaceDots(context, normalizeName(nodeName))));
+
+    String jobName;
+    if (context.getOpenLineageConfig().getJobName() != null
+        && !context.getOpenLineageConfig().getJobName().getAppendDatasetName()) {
+      // no need to append output dataset name
+      jobName = normalizeName(jobNameBuilder.toString());
+    } else {
+      // append output dataset as job suffix
+      jobNameBuilder.append(
+          getJobSuffix(context)
+              .map(
+                  suffix ->
+                      JOB_NAME_PARTS_SEPARATOR
+                          + suffix.replace(JOB_NAME_PARTS_SEPARATOR, INNER_SEPARATOR))
+              .orElse(""));
```

Comment:
If particular builder/visitor does not implement the interface, the dataset name won't be added as suffix? Would be nice to document that more visibly

Reply:
Any particular place in your mind where the doc would fit?

Reply:
https://openlineage.io/docs/integrations/spark/extending seems to be the best place